### PR TITLE
JS: Sanitizing html-escaping switch-statements

### PR DESCRIPTION
**Pull request by erik-krogh**
Gets a TN for CVE-2019-10771. 

This PR does 2 things. 

1) Generalizes barrier guards such that the test can be stored in a variable. For example: 
```JavaScript
var cond = someTest(x); // sanitizes x
if (cond) {
  // x is sanitized here.
}
```

2) Recognizes a HTML-escaping switch-statement, and sanitizes all accesses to the variable used in the switch.   
E.g. all accesses to `ch` is marked as a barrier in the below: 
```JavaScript
var buf = "";
var i = 0;
while (i < s.length) {
  ch = s[i++];
  switch (ch) {
    case "&":
      buf += "&amp;";
      break;
    case "<":
      buf += "&lt;";
      break;
    case '"':
      buf += "&quot;";
      break;
    default:
      buf += ch;
      break;
  }
}
```

Change is covered by previous change-note. 

[Smoke test evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1590661064370) shows that performance got better ¯\\\_(ツ)\_/¯

TODO:
- [ ] larger evaluation

### CODE REVIEW

**Comment by asgerf** [comment]
Regarding the switch handling, I believe this is implemented by https://github.com/github/codeql/pull/3415 as well. Could you check whether that PR is enough to get the TN?

**Comment by erik-krogh** [comment]
> Regarding the switch handling, I believe this is implemented by #3415 as well. Could you check whether that PR is enough to get the TN?

That PR is not enough to get a TN on its own, it needs an XSS specific sanitizer on top. 

I could make a generalization of this PR based on `MembershipCandidate`, but I think that might be too general, as that could also match e.g. a regexp test that tests for presence of the HTML-specific chars.

File: javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll [hunk 1]
```diff
@@ -73,6 +73,73 @@ module Shared {
...
+
+  /**
+   * A sanitizer guard that checks for the existence of HTML chars in a string.
+   * E.g. `/["'&<>]/.exec(str)`.
+   */
+  class ContainsHTMLGuard extends SanitizerGuard, DataFlow::MethodCallNode {
+    DataFlow::RegExpCreationNode regExp;
+
+    ContainsHTMLGuard() {
+      this.getMethodName() = ["test", "exec"] and
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll:85]: (Seeing this convinced me that I need to revive my old branch with `StringOps::RegExpTest`.)

File: javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll [hunk 2]
```diff
@@ -73,6 +73,73 @@ module Shared {
...
+        or
+        exists(DataFlow::PropRead read | exists(read.getPropertyNameExpr()) |
+          read.flowsToExpr(switch.getExpr()) and
+          str = read.getBase().asExpr()
+        )
+      )
+    )
+  }
+
+  private import semmle.javascript.dataflow.internal.AccessPaths as Paths
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll:123]: Hm, we're not supposed to import internal packages in our queries (and this `.qll` file counts as a query in that sense). Have you checked if the `GlobalAccessPaths.qll` can be used instead? (which, despite the filename, also builds local access paths).

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 3]
```diff
@@ -438,7 +438,7 @@ private predicate barrierGuardBlocksNode(BarrierGuardNode guard, DataFlow::Node
   barrierGuardIsRelevant(guard) and
   exists(AccessPath p, BasicBlock bb, ConditionGuardNode cond, boolean outcome |
     nd = DataFlow::valueNode(p.getAnInstanceIn(bb)) and
-    guard.getEnclosingExpr() = cond.getTest() and
+    (guard.getEnclosingExpr() = cond.getTest() or guard = cond.getTest().flow().getALocalSource()) and
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/dataflow/Configuration.qll:441]: I tried your example, and we didn't falsely block flow in that example (not sure why). 

I changed it to use `getImmediatePredecessor()` anyway.

**Comment by erik-krogh** [comment]
[A nightly/security evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1590826184944) looks good ([redo of the 5 best + 5 worst](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1590842885780)). 

It looks like this gives a 1-2% performance increase, but I'm not sure where that comes from.  
It might be because we sanitize more variables now.

**Pull request merged**